runtime/memprof.c

```diff
@@ -156,1176 +1258,983 @@ static memprof_domain_t domain_create(caml_domain_state *caml_state)
-      /* Transfer the entry to the global array. */
-      struct tracked* t = &entries_global.t[entries_global.len];
-      entries_global.len++;
-      *t = local->entries.t[idx];
-
-      if (Is_exception_result(res)) {
-        /* The allocations are cancelled because of the exception,
-           but this callback has already been called. We simulate a
-           deallocation. */
-        t->block = Val_unit;
-        t->deallocated = 1;
-      } else {
-        /* If the execution of the callback has succeeded, then we start the
-           tracking of this block..
-
-           Subtlety: we are actually writing [t->block] with an invalid
-           (uninitialized) block. This is correct because the allocation
-           and initialization happens right after returning from
-           [caml_memprof_track_young]. */
-        t->block = Val_hp(Caml_state->young_ptr + Offs_placeholder(t->block));
-
-        /* We make sure that the action pending flag is not set
-           systematically, which is to be expected, since we created
-           a new block in the global entry array, but this new block
-           does not need promotion or deallocationc callback. */
-        if (callback_idx == entries_global.len - 1)
-          callback_idx = entries_global.len;
+  /* If profiling hasn't been stopped and restarted by these
+   * callbacks, offset entries for these sampled allocations may still
+   * exist, and will be in the thread's entry table. If one of the
+   * callbacks has raised an exception, these entries should be
+   * deleted (or marked as deallocated if the allocation callback
+   * ran). Otherwise, they must be updated to point to the blocks
+   * which will now be allocated. */
+  if (!restarted) {
+    size_t i = 0;
+    if (entries->live >= new_entries) /* cope with deleted entries */
+      i = entries->live - new_entries;
+    while (i < entries->live) {
+      entry_t e = &entries->t[i];
+      if (e->offset) { /* an entry we just created */
+        if (cancelled) {
+          /* If we ran the allocation callback, keep the entry but
+           * fake de-allocation. */
+          if ((e->callbacks & CB_MASK(CB_ALLOC)) &&
+              ((e->callbacks & CB_MASK(CB_DEALLOC)) == 0)) {
+            e->offset = false;
+            e->block = Val_unit;
+            e->deallocated = 1;
+            if (i < entries->next) entries->next = i;
+          } else {
+            mark_deleted(entries, i);
+          }
+        } else {
+          e->block = Val_hp(Caml_state->young_ptr + e->block);
+          e->offset = false;
+        }
       }
+      ++i;
+    }
+
+    if (cancelled) {
+      /* evict anything we just deleted */
+      entries_evict(entries, false);
     }
-    mark_deleted(&local->entries, idx);
+
+    /* There are now no outstanding allocation callbacks in the
+     * thread's entries table. Transfer the whole thing to the
+     * domain. */
+    entries_transfer(entries, &domain->entries);
```

Comment:
If one of the callback raises an exception, then the table may contain deferred entries that we should not transfer.

```suggestion
    else {
      /* There are now no outstanding allocation callbacks in the
       * thread's entries table. Transfer the whole thing to the
       * domain. */
      entries_transfer(entries, &domain->entries);
    }
```